### PR TITLE
Spliterators and primitive streams

### PR DESCRIPTION
**Pull request by techsy730**
Here it is, spliterators and primitive stream methods for collections.
Also fill in some other Java 8 methods.

Overview of the change: https://docs.google.com/document/d/1yMf5LSajEQ4i4odnwWrXDPPl7OmjX1Fj2MpHzhCAB5c/edit?usp=sharing

This change addresses https://github.com/vigna/fastutil/issues/116, https://github.com/vigna/fastutil/issues/115, and https://github.com/vigna/fastutil/issues/98

IntArrayList.intStream().sum() is about 3.5 times faster then java.util.ArrayList<Integer>.stream().mapToInt(Integer::intValue).sum()
(1ns per element for ArrayList<Integer>, .28ns per element for IntArrayList)

If you think the changes are substantial enough to warrant a bump to 9.0.0, may I suggest that we require Java 9 for that so that we can take advantage of its high performance methods for arrays and index checks?

I am not content with how the disambiguation methods (such as forEachRemaining(...fastutil.ints.IntConsumer) delegates to forEachRemaining(java.util.function.IntConsumer)) turned out. But without this, calling forEachRemaining(...fastutil.ints.IntConsumer) would cause a compile-time error as it would be ambiguous between java.util.function.IntConsumer and java.util.function.Consumer<Integer>. If you don't think this disambiguation is worth it (for example, you wanted to deprecate ...fastutil.ints.IntConsumer), then I can remove it and the cruft associated with it.

Apologies for doubling the size of TODO. I do plan on getting to the items I listed in it, but I didn't want to block any longer on this.

### CODE REVIEW

**Comment by szarnekow** [comment]
> require Java 9 for that so that we can take advantage of its high performance methods for arrays and index checks

Out of curiosity: Which methods are you referring to?

**Comment by vigna** [comment]
So. This is an *enormous* amount of code that will require time to be reviewed. There's even a class (C) Google (huh?).

I think you need to write a document explaining the design choices and how they were implemented. I cannot navigate blindly through all of this code.

**Comment by vigna** [comment]
And, how much of the code is covered by unit tests?

**Comment by vigna** [comment]
I see Objects.checkIndex() is intrinsified in Java 9 (the other check methods do not—bummer). Thank you for the tip (and for all the code).

**Comment by simlu** [comment]
Wow, this looks terrific! Just did a quick test and for what we're doing the speed difference is significant. Looking forward to having this released (after a thorough review)!

**Comment by techsy730** [comment]
Sure, I'll write up a document explaining the new classes and methods and why I coded them as I did and link it here, since yeah, there is a lot of code to go on with little context as to why it is the way it is.

As for unit test coverage, I made sure that almost all of it was covered by tests. Either by unit tests or the runnable main tests. I am afraid I don't have hard code coverage numbers at the moment, but I can probably get those if you really want hard coverage. But I did make sure to test all Spliterators, forEachRemainings, tryAdvances, and trySplits (by way of their usage through Streams)

**Comment by techsy730** [comment]
Also, if the "Copyright Google" is unnerving, I can change that part of the blurb to be under my name (the email is set to my google.com address, so it is still credited to google). Google isn't that strict on copyright.
Also, no matter what, this will still be under Apache 2.0, same as the rest of the code, so you can still do whatever you want with it.

**Comment by vigna** [comment]
Actually, no, it's not, but I would like to know why that source is there.

**Comment by techsy730** [comment]
Basically, it is because I am contributing to this as a Google employee, and as such the author for copyright purposes is Google. That's just our policy for open source contributions "on company time".

Again, it can either by under "Google LLC" or "C. Sean Young", whichever you would be more comfortable with. The exact wording of the header is not important.

**Comment by techsy730** [comment]
> > require Java 9 for that so that we can take advantage of its high performance methods for arrays and index checks
> 
> Out of curiosity: Which methods are you referring to?

For Array methods

[java.util.Arrays.equals(T[], fromIndex, toIndex, T[], fromIndex, toIndex)](https://docs.oracle.com/javase/9/docs/api/java/util/Arrays.html#equals-java.lang.Object:A-int-int-java.lang.Object:A-int-int-)
and
[java.util.Arrays.compare(T[], fromIndex, toIndex, T[], fromIndex, toIndex)](https://docs.oracle.com/javase/9/docs/api/java/util/Arrays.html#compare-T:A-T:A-)

Of which the primitive versions are vectorized, which is much faster then we can do with "regular" Java code.

For Index checking

[java.util.Objects.checkIndex(index, Length)](https://docs.oracle.com/javase/9/docs/api/java/util/Objects.html#checkIndex-int-int-), checkFromToIndex​(fromIndex, toIndex, length), and checkFromIndexSize​(fromIndex, size, length).

Currently only one of them is intrinsicified (I think the basic checkIndex one), but possibly in the future more of them can be.

**Comment by techsy730** [comment]
A design overview doc is incoming; just putting on the finishing touches to it

**Comment by vigna** [comment]
Well, historically contributors of a significant amount of code have been signing a copyright release form. It would have been a good idea to tell me in advance you couldn't do that.

**Comment by techsy730** [comment]
Oh, I'm sorry. I didn't realize there was a process for that. Let me know what I need to do and I will run it by our open source team.

**Comment by vigna** [comment]
It's not a big issue now. It was with LGPL. But let's see what we can do.

**Comment by techsy730** [comment]
Just FYI, Google doesn't require me to sign any CLA for Apache 2.0 licensed projects. If you do on your end, that is fine, I can go through that process. But I am clear to contribute to Apache 2.0 licensed projects on our side.

**Comment by techsy730** [comment]
Ok, here is the design overview for this change:
https://docs.google.com/a/google.com/document/d/e/2PACX-1vScXXgVe0MBQYxLsxd8csFRhxE6nKF4BHb_9D_NiIKIh0LkMrxN0QVF0Dp6Dbs24M24Deikp1XfoezM/pub

Apologies for the requirement to sign in; not sure how to turn that off.

**Comment by simlu** [comment]
Yeah, getting a perm denied...

**Comment by techsy730** [comment]
Ok, let's try this
https://docs.google.com/document/d/1yMf5LSajEQ4i4odnwWrXDPPl7OmjX1Fj2MpHzhCAB5c/edit?usp=sharing

Feel free to leave comments. I've made notes about design choices I am not content about and am willing to consider alternatives.

**Comment by techsy730** [comment]
I think something I am going to try is on different branches, do those alternatives mentioned in the doc, and see if it makes the API simpler and less hacky.

**Comment by techsy730** [comment]
Please don't merge the change as it; I think I have a way to get rid of the awful ImplementsSpliterator hack

**Comment by techsy730** [comment]
I think I have a way to get rid of the awful ImplementsSpliterator hack, if you want to wait for that.

**Comment by vigna** [comment]
OK. It'll take time to review and merge this in, so take your time.

**Comment by techsy730** [comment]
Ok, here is the change I am proposing to get rid of ImplementsSpliterator, and also makes the interface methods a lot cleaner:
https://github.com/techsy730/fastutil/pull/1/files

If you think this is cleaner, then I will merge this into this branch and thus pull request.

**Comment by techsy730** [comment]
I've gone ahead an merged in the change to remove ImplementsSpliterator

Major changes:
- Iterable.ImplementsSpliterator and all sub-interfaces have been removed
- The default implementations of spliterator and intSpliterator have simplified greatly
- New methods and classes in Spliterators, one wrapping an Iterator as a Spliterator, the other wrapping a Spliterator as a Iterator. These methods are analogous to java.util.Spliterators.spliterator and iterator

The design document has been updated accordingly.

**Comment by benmccann** [comment]
@techsy730 your PR will need to be rebased. I'm not affiliated with the project, but I wonder if it might be possible to split it into multiple PRs to make it easier to review? I'm very thankful for the work you've done here and excited to see it make its way in

**Comment by vigna** [comment]
BTW, I didn't forget you... just swamped. I'll have a window to look at this in July.

**Comment by techsy730** [comment]
@benmccann It is possible, but it would be tricky because of how interconnected the commit history is. Furthermore, the Java Spliterator is pretty pervasive; it is hard to implement one part without having the other parts implemented too.
Perhaps I can do something like
Add primitive spliterator interfaces + codegen changes; add primitive spliterator() methods and utility classes; add primitive stream methods.
But would that make it be any cleaner? Doing it like this will result in PRs with a large amount of unused code until all of them go in.

Something I will absolutely look at doing is to split off the List sort method changes into their own PR; they can stand alone.

**Comment by techsy730** [comment]
Ok, rebased to HEAD, and split off the List.sort and Arrays sorting methods into their own pull commit; https://github.com/vigna/fastutil/pull/130

**Comment by vigna** [comment]
I'd release 8.3.0 as it is now. I'll try to go through these changes, even if I don't know if I can do it before September. At that point we could bump to Java >= 9 and the release to 9.0.0. It's a lot of code.

Which are the high performance methods for arrays and index checks?

I'm trying to remember why IntConsumer exists in the first place. It was a long time ago. LOL.

**Comment by techsy730** [comment]
@vigna The high performance methods in Java 9 I am talking about are
java.util.Objects.checkFromIndexSize, checkFromToIndex, and checkIndex (only one of these is currently intrinsicified, but we should still use all of them in case others get it in the future)
java.util.Arrays.equals (overloads accepting bounds were added in Java 9)
java.util.Arrays.compareTo

**Comment by vigna** [comment]
Ah ok. I was guessing that but since only one was intrinsified I thought I was missing something. Certainly starting to use them is a good strategy.

File: CHANGES [hunk 1]
```diff
@@ -17,6 +32,7 @@
   easily by static imports. The old versions have been deprecated.
 
 - New methods to unwrap iterators into big arrays.
+>>>>>>> mainline/master
```
Comment (benmccann) [inline CHANGES]: is this a merge conflict?

**Comment by vigna** [comment]
I'll finally get time to put this into the release. I'd stay for Java 8 at least this round. You'll hear from me soon...

**Comment by vigna** [comment]
"Soon" is a month ago but yes, you'll hear from me :).

**Comment by vigna** [comment]
And it took one year. Sorry, work, COVID-19, and the rest.

This PR is the largest contribution ever entering fastutil. I've read your support documentation and worked a bit through the code. The contribution is clearly of excellent quality and introduces a much-needed set of features to fastutil.

In terms of code style, the only differences I can see are for-loops with ++i instead of i++ and braces around one-line body of conditionals, etc. Not a big issue, I can do a pass by myself to fix those.

Two files have a Copyright by Google. In general, I've always asked contributors to fastutil to release the copyright to me, GNU style. Indeed 90% of the code you added is in existing files with my copyright, so I think it should not be a big problem. Do you need to ask your employer maybe? 

I invite you to put an @author tag in all files in which you contributed code.

Finally, there is a dire need for a section in overview.html that explains the basics of the design (as you did in the document above). Would you be so kind to write it? Even a rough draft is OK.

And of course, in one year a few conflicts arose. But those should be easy to fix (I hope).

**Comment by vigna** [comment]
Oh, a last question: how much coverage of the new code did you get with your unit tests?

**Comment by techsy730** [comment]
Thanks for taking a look. :)
I am busy with work right now, but I should be able to address your comments during the weekend.

**Comment by vigna** [comment]
Well take your time. I took 18 months... 🙈

**Comment by techsy730** [comment]
Ok, I fixed the merge conflicts, updated the copyright headers, added a section to overview.html about the spliterators and stream methods.

Getting code coverage has been a bit tricky as the tests are spread across the unit tests and the legacy static main method tests (which I think I have a way to integrate into the unit tests, I will make a new PR with that).
I'll post another note with coverage numbers I can get.

**Comment by techsy730** [comment]
Oh, something I would like to add before this being accepted are toList and toSet methods on IntArrayList, IntArraySet, and IntOpenHashSet (and their corresponding Long and Double versions), that take a primitive stream and terminates it collecting it into a list/set. (It won't return a Collector as there is no primitive Collector equivalent in the Java API, it would just have to take the stream)

**Comment by vigna** [comment]
That's amazing!

**Comment by techsy730** [comment]
Ok that is ArrayList.toList() and OpenHashSet.toSet() collector-ish methods are done (and true Collectors for the object based versions)

However, I'm on the fence about ArraySet:
First, I need to decide whether to deduplicate or not.
If I do, I have three options:
1. Call IntStream.distinct, which currently boxes and unboxes into a object based map, which goes against the "ethos" of fastutils
2. First accumulate into OpenHashSet and then copy into ArraySet after it is done, which preserves linear time performance and no boxing, but does have an extra copy step at the end
3. Accumulate into ArraySet directly and take a quadratic loop (linear time search every insert).

If I were to not deduplicate, it would be easy to create invalid state ArraySets.
While asking for uniqueness from the constructors is easy enough to ask callers to verify, in Streams it is not because of the arbitrary operations you can do on stream pipelines.

I'm wondering if ArraySet should even get a collector because ArraySet's with lots of elements are a big performance no-no.

**Comment by vigna** [comment]
Er... why not accumulate in an array, radix-sort it, and deduplicate it in a single pass? That's basically linear, and basically with no additional space (you can use any sort, in fact, like, the platform sort).

**Comment by techsy730** [comment]
Good idea with the sorting. That way I just need to ensure that no two adjacent items are equal.
I will still have to dump to a hash set for large Object streams (as not all Objects are Comparable), but for primitives a sort is great.

As for code coverage, most of my "coverage" was in the form of additions to the main method "stress tests". While coverage using those is pretty good (the version of Eclipse I am using doesn't give delta code coverage over new code, so I can't give a hard number there), I think I should also make them proper unit tests (since we are moving that direction anyways).

**Comment by incaseoftrouble** [comment]
sorry for deviating from the original issue a bit: What definitively should also be in there eventually is property based testing. We can simply use the java collections as reference, generate a stream of random mutations, and check for fidelity.

**Comment by vigna** [comment]
Yes, equal behavior with java.util is what the messy main() tests are trying to achieve...

**Comment by vigna** [comment]
I checked a bit around and @implNote etc. are non-standard. They must be enabled with an option in the tool javadoc and might change in the future. I would refrain from using them in the Javadoc documentation—just add a header "Implementation note" if you need it.

**Comment by techsy730** [comment]
While @implNote and friends are non-standard (and indeed the docket warns you as such), the ability to make your own Javadocs tags is standard.
https://docs.oracle.com/javase/1.5.0/docs/tooldocs/windows/javadoc.html#options (search for "-tag tagname").

That being said, since the Javadocs tool recognized @implSpec and friends as "special", would you rather rename them to something that isn't recognized? Or would you rather refrain from using this Javadocs feature at all?

**Comment by vigna** [comment]
Ok. So essentially as long as we put in build.xml all the relevant -tag options those tags will always work (even if the jdk stops using them). So let's go on. Can you modify the javadoc target of build.xml so they're processed correctly?

**Comment by techsy730** [comment]
Looks like I already did update `build.xml`

And I will get on the spliterator and stream unit tests this weekend, as well as clean up the ArraySet collector methods.

**Comment by techsy730** [comment]
I've got the merge conflict worked out and have a few more tests for the unit test classes, and then I should be ready for review.

**Comment by techsy730** [comment]
Ok, for some code coverage numbers:

- `IntArrayList.Spliterator` has 71.3% coverage. The only thing untested was the `tryAdvance`, as the stream framework usually goes through `forEachRemaining`)
- `IntOpenHashSet.Spliterator` has 51.0% coverage. The only things untested were `skip` (it would be ugly to test this on a spliterator without an ordering assurance) and `tryAdvance` (same reason as above)
- `IntSpliterators` has 51.4% coverage (Most of the uncovered parts are the widening spliterators, like ShortSpliterator -> IntSpliterator wrapper, or the abstract classes)

**Comment by techsy730** [comment]
On second thought, I should add the spliterator tests to the short or byte classes. That way the widening logic to ints can be covered.

(Unfortunately my dev machine just died. I have a spare but setting things up on it will take time)

**Comment by vigna** [comment]
Hey, how are things going with the new machine? :)

**Comment by techsy730** [comment]
I'm more or less comfortable with the state as is being reviewed. Of course I need to fix that merge conflict too.

Machine still in the shop. Haven't had time to setup my other machine with this project yet. I can do that this weekend.

**Comment by techsy730** [comment]
Ok, merge conflicts fixed and Java 8 methods added to ImmutableList.

**Comment by vigna** [comment]
Line 146 of BigList.drv seems a bug to me. You should use length(a), not a.length—the latter is just the number of segments. Right?

**Comment by vigna** [comment]
I merged, but could you add a couple of examples of usage at the end of "an example" section at the end of the overview? I think that would really be helpful.

**Pull request merged**